src/test/ensemble_tests.cpp

```diff
@@ -1834,3 +1834,124 @@ TEST_F(EnsembleFlowTest, ErrorHandlingSkipsDeferredNodesExecutionIfExecutionFail
+    ASSERT_TRUE(status.ok()) << status.string();
+    std::vector<NodeInfo> infoNew{
+        {NodeKind::ENTRY, "request"},
+        {NodeKind::DL, "dummy_node", "newDummy"},
+        {NodeKind::EXIT, "response"},
+    };
+    status = pd.reload(managerWithDummyModel, infoNew, connections);
+    EXPECT_TRUE(status.ok()) << status.string();
+}
+
+TEST_F(EnsembleFlowTest, ReloadPipelineDefinitionWithNewNonExistingModelNameShouldFail) {
+    ConstructorEnabledModelManager managerWithDummyModel;
+    managerWithDummyModel.reloadModelWithVersions(config);
+
+    const std::string pipelineName = "originalName";
+    std::vector<NodeInfo> info{
+        {NodeKind::ENTRY, "request"},
+        {NodeKind::DL, "dummy_node", "dummy"},
+        {NodeKind::EXIT, "response"},
+    };
+    std::unordered_map<std::string, std::unordered_map<std::string, InputPairs>> connections;
+    // request O--------->O dummy node
+    connections["dummy_node"] = {
+        {"request", {{"pipeline_input_name", "b"}}}};
+    // dummy node O--------->O dummy node 1
+    connections["response"] = {
+        {"dummy_node", {{"a", "pipeline_output_name"}}}};
+    PipelineDefinition pd(pipelineName, info, connections);
+    auto status = pd.validate(managerWithDummyModel);
+    ASSERT_TRUE(status.ok());
+
+    ASSERT_TRUE(status.ok()) << status.string();
+    std::vector<NodeInfo> infoNew{
+        {NodeKind::ENTRY, "request"},
+        {NodeKind::DL, "dummy_node", "newDummy"},
+        {NodeKind::EXIT, "response"},
+    };
+    status = pd.reload(managerWithDummyModel, infoNew, connections);
+    EXPECT_EQ(status, ovms::StatusCode::MODEL_NAME_MISSING) << status.string();
+}
+
+TEST_F(EnsembleFlowTest, ReloadPipelineDefinitionWithAllModelVersionsRetiredShouldFail) {
+    ConstructorEnabledModelManager managerWithDummyModel;
+    managerWithDummyModel.reloadModelWithVersions(config);
+
+    const std::string pipelineName = "originalName";
+    std::vector<NodeInfo> info{
+        {NodeKind::ENTRY, "request"},
+        {NodeKind::DL, "dummy_node", "dummy"},
+        {NodeKind::EXIT, "response"},
+    };
+    std::unordered_map<std::string, std::unordered_map<std::string, InputPairs>> connections;
+    // request O--------->O dummy node
+    connections["dummy_node"] = {
+        {"request", {{"pipeline_input_name", "b"}}}};
+    // dummy node O--------->O dummy node 1
+    connections["response"] = {
+        {"dummy_node", {{"a", "pipeline_output_name"}}}};
+    PipelineDefinition pd(pipelineName, info, connections);
+    auto status = pd.validate(managerWithDummyModel);
+    ASSERT_TRUE(status.ok()) << status.string();
+    managerWithDummyModel.findModelByName("dummy")->retireAllVersions();
+
+    status = pd.reload(managerWithDummyModel, info, connections);
+    EXPECT_EQ(status, ovms::StatusCode::MODEL_NAME_MISSING) << status.string();
+}
+
+TEST_F(EnsembleFlowTest, ReloadPipelineDefinitionWhen1ModelVersionBecomesAvailableShouldPass) {
```

Comment:
Maybe add one more test that ensures creating pipeline out of definition does wait for reload to happen?

Reply:
For this I need PipelineStatus functional.